### PR TITLE
Gravatar image update

### LINKED ISSUE #3228: Change gravatar photo in app
Currently there is no way to change the photo within the app.

### PR DESCRIPTION
**Pull request by hypest**
Fixes #3228 

To test: Be logged in a .com site and head over to the profile screen ("me"). Tap the current Gravatar picture, select of capture a new one, crop it to your liking and finish. The app should then take some seconds to update the Gravatar remotely and refresh the screen, displaying the newly set Gravatar.

Known shortcoming: the current picture frame in the profile screen has no indication that it is now tappable.

This solution brings in okHttp (and okio).

Needs review: @bummytime (you may want to defer to @maxme for the network layer)

### CODE REVIEW

**Comment by bummytime** [comment]
Nice work @hypest. I tested this feature on a wide variety of devices and API levels and it seems pretty solid overall. The upload did hang for me a few times on a Samsung S6 on 4G, however I was not able to reliably repro the issue. If you want, I can try and flesh it out a little more but I would say it may have been the carrier network I was on at the time.

> Known shortcoming: the current picture frame in the profile screen has no indication that it is now tappable.

I agree this is a shortcoming. Perhaps we could put some kind of graphical affordance on the gravitar to suggest it is tappable... e.g. an animation or tooltip. This shortcoming may not be the worst thing on earth, but I wouldn't want this great feature to go unused either ;-)

Code-wise, everything looks in good order, however I would feel more comfortable if @maxme gave it his stamp of approval. :grinning:

**Comment by hypest** [comment]
Thanks for the review @bummytime ! Glad you liked it ;)

I'll try it myself with an S6 as well but yeah, please give it a another go on your end.

Let's wait for @maxme 's review as well and I'll attack the shortcoming in the meantime.

Thanks!

**Comment by maxme** [comment]
> Known shortcoming: the current picture frame in the profile screen has no indication that it is now tappable.

Yes, maybe we could add a "Change Profile Picture" entry in "My Profile" - cc @mattmiklic 

This is a great addition, our users were requesting this feature for a long time. 

Some comments:
* My first experience went badly because I had the Camera permission disabled (Android 6). We should display a Toast message if the user refused the Camera permission previously (currently, it asks the user to choose between camera or doc, if camera is chosen nothing happen).
* In landscape, margins seems too big around the profile picture: https://cloudup.com/co7Wx0UFjWo
* Previous profile pic must be cached somewhere, we should invalidate this cached picture. After a rotation (or any other onPause/onResume) the new profile picture is lost, old one comes back - See: https://cloudup.com/cuAidxtcrZA

File: WordPress/src/main/res/layout/me_fragment.xml [hunk 1]
```diff
@@ -14,16 +14,35 @@
...
-            <org.wordpress.android.widgets.WPNetworkImageView
-                android:id="@+id/me_avatar"
-                android:layout_width="match_parent"
-                android:layout_height="match_parent" />
-
+            <FrameLayout
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:padding="@dimen/content_margin"
+                android:background="?android:attr/selectableItemBackgroundBorderless" tools:targetApi="lollipop">
```
Comment (maxme) [inline WordPress/src/main/res/layout/me_fragment.xml]: App crashes on API < 21 here.

**Comment by mattmiklic** [comment]
> Yes, maybe we could add a "Change Profile Picture" entry in "My Profile" - cc @mattmiklic

This is probably the easiest way to do this. Since there's no hover state like on the web, it's a little harder to design an icon or some visual clue on the gravatar itself that it's tappable. I'd be OK with making the gravatar tappable, without a visual clue (think of it as a shortcut for those in the know) along with a dedicated text link like @maxme described here.

**Comment by hypest** [comment]
Thanks for the review @bummytime  and @maxme !

I've addressed all but one of the issues Maxime brought up so feel free to resume the review.

The only issue not addressed yet is the one where the image is cached and is shown again on rotation. I'm currently stuck at the Gravatar side where there seems to be a cache in place and needs some variable number of seconds to invalidate.

**Comment by maxme** [comment]
screenshot-2016-04-19_16 35 45 658

I'm not sold on the tooltip - what do you think @mattmiklic? @hypest what do you think of dropping the tooltip and adding a text link `Change Profile Pic` in "My profile"?

If we keep the tooltip can we show it only for users who upgrade to version `5.4` (= not new user and using version 5.4)? 


Confirmed that:
- [x] Toast is shown when the user disabled camera (we could do better if the user denies one of the two permissions, maybe fill a new low priority issue)
- [x] Better margins

**Comment by mattmiklic** [comment]
> I'm not sold on the tooltip - what do you think @mattmiklic? @hypest what do you think of dropping the tooltip and adding a text link Change Profile Pic in "My profile"?

Agreed that no visual indication is needed on the Gravatar itself if we add a link to "Change profile picture" under my profile.

**Comment by hypest** [comment]
> > I'm not sold on the tooltip - what do you think @mattmiklic? @hypest what do you think of dropping the tooltip and adding a text link Change Profile Pic in "My profile"?

> Agreed that no visual indication is needed on the Gravatar itself if we add a link to "Change profile picture" under my profile.

Hmm, I think we're talking about a couple of different things here. The tooltip is one thing and what/where the user taps to start the gravatar change flow is another:
* Tapping the picture to change it is the most natural thing to do, like other similar apps do. Besides, that's the flow we agreed with @mattmiklic from the beginning. @mattmiklic , not clear from your comment above but: Are you suggesting removing the functionality of tapping the image to start the flow?
* Assuming you guys advocate for removing the tapability of the image and moving that functionality into the `My Profile` screen, I'd say that that's an indication of a design problem: Like, why is the image displayed on one screen but the user needs to navigate into another to change it? Doesn't make much sense to me.
* Assuming you guys advocate _adding_ an entry point into "My Profile" as well, I'd say that this would introduce a second entry point and that would be confusing IMHO. Better to keep only one entry point.

**Comment by mattmiklic** [comment]
> Are you suggesting removing the functionality of tapping the image to start the flow?

No. I think tapping the image to upload the gravatar is the best option. However, I do not think an icon, tooltip, or any other visual accompaniment is necessary.

If you look at the example of Instagram's profile page, simply tapping on your profile photo opens a menu to replace it. If that interaction is discoverable enough for Instagram, I feel fairly confident that our users will be able to understand it, too. 

This is the way gravatar uploading has been implemented in WordPress-iOS, it's in the internal beta build now. Tapping the gravatar initiates a small animation and opens the image picker.

I don't think a secondary text link within the my profile section is necessary.

**Comment by rickybanister** [comment]
I agree 💯 percent.

After seeing infants use smart phones, I'm pretty positive that tapping the avatar is findable enough.

**Comment by hypest** [comment]
Cool! Tap-the-gravatar-to-change it is then!

Not arguing here but, just to clear up something small: The reason of the tooltip's addition/proposal. The current production app has a non-tappable gravatar. Tapping it makes nothing happen. It's safe to assume that current users have already found that out.
But, the new version will make the image tappable. The tooltip is only shown *once* and never again. It is there to let those users know that they can now trust their instinct again and tap the image to change it.
When we are confident enough that we informed our current userbase enough (say, in 2 or 3 releases down the road), we can retire the tooltip altogether.

The PR at the moment still has the tooltip commit included so, let's decide on this by answering this question ( @mattmiklic and/or @rickybanister ):
Shall I remove the one-time tooltip completely?

OK, with the tooltip related thing out of the way, lemme share a video of an "pencil" implementation I've been working since yesterday. https://cloudup.com/iZHgmiT6SnB
The rationale of the "pencil" is different to the tooltip of course. The pencil is meant to always appear there, similar to what Facebook, Twitter and Slack are doing with their "camera" icon on their Android apps which is always overlaid on the avatar pic.
Not pushing here though. I'm only sharing this for completeness and, well, because it has a nice little animation :)

**Comment by rickybanister** [comment]
A one-time tooltip seems more reasonable than an all-the-time solution, but I don't have a big stake in either decision.

**Comment by mattmiklic** [comment]
> A one-time tooltip seems more reasonable than an all-the-time solution

Agreed, I'd be fine with a one-time hint, conceptually similar to the way we've called out new features in the editor and stats before.

I'd also be fine with, as a later iteration, testing whether adding a permanently-visible icon would have a significant impact on the number of users finding the feature.

**Comment by kwonye** [comment]
> I'd also be fine with, as a later iteration, testing whether adding a permanently-visible icon would have a significant impact on the number of users finding the feature.

Aaannnddddd thanks to @maxme's suggestion, this will be our first Optimizely A/B test! 🎉 #3988

**Comment by maxme** [comment]
Ok, so the only thing blocking the merge is a small bug with the tooltip being shown when a user logins with a self hosted account:

screenshot-2016-04-28_08 30 48 338

And a visual bug (nitpick, not even sure we should fix it):

https://cloudup.com/cJzkaHKYRk4

**Comment by hypest** [comment]
@bummytime , @maxme , I've pushed the fixes for the rest of the issues brought up (thx @maxme !) and I think this PR is ready for another round of review and hopefully merge. Cheers!

**Comment by nbradbury** [comment]
Hey all, hope you don't mind me jumping in here, but I've been wanting this feature for three years now so I just had to give it a spin. It works well, but I have a few concerns:

* When I tap to change the profile pic, I see an action sheet that shows "Camera" and "Documents" as choices (<a href="https://cloudup.com/cKg0wSYVmoD">screenshot here</a>). That's really confusing - it doesn't make sense that you have to select "Documents" to choose from your photo library. Some other apps that enable changing the profile pic show a simple popup menu ("Take photo" / "Choose existing photo"), and I think that would work much better.

* The cropping library is a bit weird. You can keep scrolling the zoom percentage without anything changing (<a href="https://cloudup.com/cDNn2lSVn52">demo video</a>).

* This single feature adds three new dependencies to `build.gradle`, which I think will come back to bite us. Can we implement this without requiring all of them? I think we could drop the cropping dependency and simply crop the photo automatically (center crop from the top). And rather than rely on a tooltip, we could apply a subtle animation to the profile pic that mimics the pressed state to draw attention to it.

**Comment by maxme** [comment]
> Can we implement this without requiring all of them? I think we could drop the cropping dependency and simply crop the photo automatically (center crop from the top).

I don't like adding new dependencies, but I think the cropping feature is a must have for a profile pic. If I want to use any pic on my phone with my face on it, center crop will always produce "funny" results.

Already wrote it in this thread but i'd prefer nothing at all than a tooltip ;).

**Comment by hypest** [comment]
Hey @nbradbury , thanks for chiming in on this! Please see my comments below:

> That's really confusing - it doesn't make sense that you have to select "Documents" to choose from your photo library. Some other apps that enable changing the profile pic show a simple popup menu ("Take photo" / "Choose existing photo"), and I think that would work much better.

That's a fair point! I could use the `ACTION_PICK` Intent action (which does bring up "Gallery/Photos" in place of "Documents") instead of the `ACTION_GET_CONTENT` but unfortunately, the former should be _considered_ deprecated. See Dianne Hackborn's answer here: http://stackoverflow.com/a/6486827. It's unfortunate that the system doesn't offer a more suitable icon to denote that we're picking up "image" documents when using ACTION_GET_CONTENT.

> The cropping library is a bit weird. You can keep scrolling the zoom percentage without anything changing

Indeed, I should probably open an issue with the library at their GitHub project. Thanks for reporting!

> This single feature adds three new dependencies to build.gradle, which I think will come back to bite us. Can we implement this without requiring all of them? I think we could drop the cropping dependency and simply crop the photo automatically (center crop from the top).

Another fair point! To be honest, I struggled with that bit when I was implementing the feature. At the end of the day, recreating the functionality we (and most probably the users as well) wanted (that is, to be able to crop and zoom) was too much and would be a waste of resources... A library is the best route for such non-core functionality. We _can_ iterate on this though after releasing the PR if we insist.

Allow me to comment though that doing some automatic cropping is definitely *not* the way to go as a first iteration on the feature as it will require us to solve a bigger problem (to detect a face and its scale and crop around it). I don't think that the "average" picture would have the face centered and scaled in favor to just do a fixed crop. A smart crop would probably require a library on its own anyway, bringing us back to the "yet another dependency" problem.

> And rather than rely on a tooltip, we could apply a subtle animation to the profile pic that mimics the pressed state to draw attention to it.

Regarding removing the library dependency for this one, I'd say that it's probably not worth it since the lib only adds 73 methods to the count (as reported by `dexcount-gradle-plugin`).
Regarding the tooltip itself versus some other facility to draw attention, I think it's not a blocker for this PR (especially since there's been quite some discussion around it already) and we can iterate on it later (even A/B test it!).

**Comment by hypest** [comment]
Hey @nbradbury . After some consideration, I pushed a commit that actually uses the `ACTION_PICK` method. I think it makes sense to lean towards the user-friendliness side on this one. Thanks for stirring this up!

**Comment by maxme** [comment]
Looks good :shipit:  @hypest can you merge `develop` into this branch?

**Pull request merged**